packages/foundry/contracts/hooks/SwapDiscountHook/DiscountCampaignFactory.sol

```diff
@@ -0,0 +1,240 @@
+
+        // Store campaign details
+        campaignData.campaignAddress = address(discountCampaign);
+        campaignData.owner = msg.sender;
+
+        TransferHelper.safeTransferFrom(params.rewardToken, msg.sender, address(discountCampaign), params.rewardAmount);
+
+        emit CampaignCreated(
+            address(discountCampaign),
+            params.rewardAmount,
+            params.expirationTime,
+            params.coolDownPeriod,
+            params.discountAmount,
+            params.pool,
+            params.owner,
+            params.rewardToken
+        );
+
+        return address(discountCampaign);
+    }
+
+    /**
+     * @notice Update the campaign details for a specific pool.
+     * @dev Updates the campaign details if the campaign has expired and belongs to the caller.
+     * @param params The parameters for updating the discount campaign.
+     */
+    function updateCampaign(
+        CampaignParams memory params
+    ) external campaignExists(params.pool) onlyCampaignOwner(params.pool) campaignNotExpired(params.pool) {
+        validateTokenAndPool(params.pool, params.rewardToken);
+
+        address campaignAddress = discountCampaigns[params.pool].campaignAddress;
+
+        IDiscountCampaign.CampaignDetails memory campaignDetails = _prepareCampaignDetails(params);
+
+        IDiscountCampaign(campaignAddress).updateCampaignDetails(campaignDetails);
+
+        uint256 rewardAmount = params.rewardAmount;
+        uint256 expirationTime = params.expirationTime;
+        uint256 coolDownPeriod = params.coolDownPeriod;
+        uint256 discountAmount = params.discountAmount;
+        address pool = params.pool;
+        address owner = params.owner;
+        address rewardToken = params.rewardToken;
+
+        TransferHelper.safeTransferFrom(rewardToken, msg.sender, campaignAddress, rewardAmount);
+
+        emit CampaignUpdated(
+            campaignAddress,
+            rewardAmount,
+            expirationTime,
+            coolDownPeriod,
+            discountAmount,
+            pool,
+            owner,
+            rewardToken
+        );
+    }
+
+    /**
+     * @notice Prepares the campaign details for creating or updating a campaign.
+     * @dev Creates a CampaignDetails struct with the specified parameters.
+     * @param params The parameters for preparing the campaign details.
+     * @return A `CampaignDetails` struct containing the prepared campaign details.
+     */
+    function _prepareCampaignDetails(
+        CampaignParams memory params
+    ) internal view returns (IDiscountCampaign.CampaignDetails memory) {
+        return
+            IDiscountCampaign.CampaignDetails({
+                campaignID: keccak256(abi.encode(block.timestamp, params.expirationTime)),
+                rewardAmount: params.rewardAmount,
+                expirationTime: block.timestamp + params.expirationTime,
+                coolDownPeriod: params.coolDownPeriod,
+                discountRate: params.discountAmount,
```

Comment:
The amount is the rate? Would make the naming consistent (and a percentage, if that's what it is). For instance, `rebatePercentage` (as it's really a rebate, not a discount).

Generally in this system:

A "percentage" is explicitly named as such, as is represented as an 18-decimal FixedPoint value, using the corresponding library (1 = 1e18).

A "rate" is also represented as an FP value, but represents a ratio or something like that - not a "percentage" like a swap fee.

An "amount" (either raw = native token decimals, or scaled18 = 18 decimal FP) is a quantity of something, usually tokens.

Reply:
Also, all these things would need validation in production. For instance, you'd want to make sure the coolDownPeriod isn't 100 years, percentage values are properly bounded (e.g., <= 100%, or some lower maximum value). Maybe the reward amount is bounded by some minimum/maximum that makes sense, etc.

Reply:
apart from the name from discount to rebate you are suggesting changing the parameter from discountAmount to discountRate right? Since it's percentage.

Reply:
If it's a percentage, say "percentage" explicitly (e.g., `discountPercentage` or `rebatePercentage`). Rates are usually just numbers (e.g., 1.2, 1.05, 0.88) that might represent some kind of ratio or exchange rate, but isn't a percentage. If you mean a literal 0-100% percentage, as you seem to here, it's best to use "percentage" for clarity.